api/src/main/java/io/grpc/StatusOr.java

```diff
@@ -0,0 +1,40 @@
+package io.grpc;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/** Either a Status or a value. */
+public class StatusOr<T> {
+  private StatusOr() {}
+
+  public static <T> StatusOr<T> fromValue(T value) {
+    StatusOr<T> result = new StatusOr<T>();
+    result.hasValue = true;
+    result.value = checkNotNull(value, "value");
+    return result;
+  }
+
+  public static <T> StatusOr<T> fromStatus(Status status) {
+    StatusOr<T> result = new StatusOr<T>();
+    result.status  = checkNotNull(status, "status");
+    checkArgument(!status.isOk(), "cannot use OK status: %s", status);
+    result.hasValue = false;
+    return result;
+  }
+
+  public boolean hasValue() {
+    return hasValue;
+  }
+
+  public T value() {
+    return value;
+  }
+
+  public Status status() {
```

Comment:
Javadoc, and define what happens if there is a value. For example, maybe this returns an OK status.

Reply:
Done.